drivers/flash/flash_stm32_ospi.c

```diff
@@ -0,0 +1,1398 @@
+	/* Number of bytes per page */
+	uint16_t page_size;
+	int cmd_status;
+};
+
+static inline void ospi_lock_thread(const struct device *dev)
+{
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+}
+
+static inline void ospi_unlock_thread(const struct device *dev)
+{
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+
+	k_sem_give(&dev_data->sem);
+}
+
+static int ospi_send_cmd(const struct device *dev, OSPI_RegularCmdTypeDef *cmd)
+{
+	const struct flash_stm32_ospi_config *dev_cfg = dev->config;
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+	HAL_StatusTypeDef hal_ret;
+
+	LOG_DBG("Instruction 0x%x", cmd->Instruction);
+
+	dev_data->cmd_status = 0;
+
+	hal_ret = HAL_OSPI_Command(&dev_data->hospi, cmd, HAL_OSPI_TIMEOUT_DEFAULT_VALUE);
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to send OSPI instruction", hal_ret);
+		return -EIO;
+	}
+	LOG_DBG("CCR 0x%x", dev_cfg->regs->CCR);
+
+	return dev_data->cmd_status;
+}
+
+static int ospi_read_access(const struct device *dev, OSPI_RegularCmdTypeDef *cmd,
+			    uint8_t *data, size_t size)
+{
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+	HAL_StatusTypeDef hal_ret;
+
+	cmd->NbData = size;
+
+	dev_data->cmd_status = 0;
+
+	hal_ret = HAL_OSPI_Command(&dev_data->hospi, cmd, HAL_OSPI_TIMEOUT_DEFAULT_VALUE);
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to send OSPI instruction", hal_ret);
+		return -EIO;
+	}
+
+	hal_ret = HAL_OSPI_Receive_IT(&dev_data->hospi, data);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to read data", hal_ret);
+		return -EIO;
+	}
+
+	k_sem_take(&dev_data->sync, K_FOREVER);
+
+	return dev_data->cmd_status;
+}
+
+static int ospi_write_access(const struct device *dev, OSPI_RegularCmdTypeDef *cmd,
+			     const uint8_t *data, size_t size)
+{
+	const struct flash_stm32_ospi_config *dev_cfg = dev->config;
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+	HAL_StatusTypeDef hal_ret;
+
+	LOG_DBG("Instruction 0x%x", cmd->Instruction);
+
+	cmd->NbData = size;
+
+	dev_data->cmd_status = 0;
+
+	/* in OPI/STR the 3-byte AddressSize is not supported by the NOR flash */
+	if ((dev_cfg->data_mode == OSPI_OPI_MODE) &&
+		(cmd->AddressSize != HAL_OSPI_ADDRESS_32_BITS)) {
+		LOG_ERR("OSPI wr in OPI/STR mode is for 32bit address only");
+		return -EIO;
+	}
+
+	hal_ret = HAL_OSPI_Command(&dev_data->hospi, cmd, HAL_OSPI_TIMEOUT_DEFAULT_VALUE);
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to send OSPI instruction", hal_ret);
+		return -EIO;
+	}
+
+	hal_ret = HAL_OSPI_Transmit_IT(&dev_data->hospi, (uint8_t *)data);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to read data", hal_ret);
+		return -EIO;
+	}
+
+	k_sem_take(&dev_data->sync, K_FOREVER);
+
+	return dev_data->cmd_status;
+}
+
+/*
+ * Gives a OSPI_RegularCmdTypeDef with all parameters set
+ * except Instruction, Address, DummyCycles, NbData
+ */
+static OSPI_RegularCmdTypeDef ospi_prepare_cmd(uint8_t transfer_mode, uint8_t transfer_rate)
```

Comment:
Minor, but for performance reasons the struct copying could be avoided by using a pointer to the command struct. The function could fill in the values using it.